internal/pkg/agent/application/upgrade/marker_watcher_test_util.go

```diff
@@ -0,0 +1,30 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// CreateMarkerTestDir creates a temporary directory that's needed
+// for the MarkerFileWatcher to function correctly. This allows tests that
+// directly or indirectly use the MarkerFileWatcher to succeed. The
+// temporary directory is cleaned up when the test completes.
+func CreateMarkerTestDir(t *testing.T) {
+	execPath, err := os.Executable()
+	require.NoError(t, err)
+
+	testDataDir := filepath.Join(filepath.Dir(execPath), "data")
```

Comment:
Since this is only to be used in tests could you rename the file to `marker_watcher_util_test.go` or make it available only to test code in some way ?

Reply:
+1

Reply:
or just move it to the file where it's used

Reply:
> Since this is only to be used in tests could you rename the file to `marker_watcher_util_test.go` or make it available only to test code in some way ?

If I rename the file `marker_watcher_util_test.go` it only becomes available to test code in the same package.  However, I need this function in a test in another package too.  That's why my solution was to name it `marker_watcher_test_util.go`.  This is the same approach taken by [`logger.NewTesting`](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/pkg/core/logger/testing.go#L15-L26).

> or just move it to the file where it's used

It's used from two different packages, so unfortunately it can't be moved to where it's used (without duplication).

Reply:
I would rather have the duplication and break the unit test dependency between the 2 packages...
Why would this detail be needed in 2 packages? Is there more than 1 package that needs to access the marker file directly and that dependency cannot be mocked away ?

Reply:
>  Is there more than 1 package that needs to access the marker file directly and that dependency cannot be mocked away ?

The unit test outside this package that calls his helper function is [`application.TestLimitsLog`](https://github.com/elastic/elastic-agent/blob/64563943437a4703e56498ca215a2a0881e4984b/internal/pkg/agent/application/application_test.go#L50).  If I remove the call to `upgrade.CreateMarkerTestDir(t)` from that test, it fails like so:

```
=== RUN   TestLimitsLog
    application_test.go:69: 
        	Error Trace:	/Users/shaunak/development/github/elastic-agent/internal/pkg/agent/application/application_test.go:69
        	Error:      	Received unexpected error:
        	            	failed to create upgrader: failed to set watch on upgrade marker's directory [/private/var/folders/mc/qr61t2n91z3g63ffq5sy4cm00000gn/T/GoLand/data]: lstat /private/var/folders/mc/qr61t2n91z3g63ffq5sy4cm00000gn/T/GoLand/data: no such file or directory
        	Test:       	TestLimitsLog
--- FAIL: TestLimitsLog (0.00s)
```

That's because that unit test calls `Application.New`, which ends up eventually calling `upgrade.newMarkerFileWatcher()`, which is where this error is coming from because the upgrade marker's directory doesn't exist at that point.  In normal (non-test) execution, the directory exists because it's created at installation time, before Agent is run and `Application.New` is called.

Anyway, I think we can remove this dependency if I move the instantiation of the watch inside Run() as you're alluding to in https://github.com/elastic/elastic-agent/pull/3694#discussion_r1389062050.  So let me give that a shot and see if it works.

Reply:
As expected, after simplifying the implementation of `UpgradeMarkerWatcher` in 8b9d67c797848ffc33c2c3fe39bf9751b2aded64, I was able to remove this dependency as the `UpgradeMarkerWatcher` constructor no longer adds the watch on the upgrade marker's directory; that is now being done in the `Run()` method instead, which is not in the execution path of this test.